File: patterns/mqtt-state-supervision/rules-js/mqtt-state-supervision.js
const supervisionTimers = new Map();
const T_RB_SECONDS = 30;

rules.JSRule({
  name: "MQTT State Supervision",
  triggers: [triggers.ItemCommandTrigger("MySwitch")],
  execute: (event) => {
    const expectedState = String(event.receivedCommand);
    const itemName = event.itemName;

    if (supervisionTimers.has(itemName)) {
      supervisionTimers.get(itemName).cancel();
    }

    console.log(itemName + ": Ueberwachung gestartet, erwarte " + expectedState + " innerhalb " + T_RB_SECONDS + "s");

    const t = actions.ScriptExecution.createTimer(
      time.ZonedDateTime.now().plusSeconds(T_RB_SECONDS),
      () => {
        console.warn(itemName + " hat den Zustand " + expectedState + " nicht erreicht");
        // z. B. Benachrichtigung senden
        supervisionTimers.delete(itemName);
      }
    );
    supervisionTimers.set(itemName, t);
  }
});

rules.JSRule({
  name: "MQTT State Update",
  triggers: [triggers.ItemStateChangeTrigger("MySwitch")],
  execute: (event) => {
    if (supervisionTimers.has(event.itemName)) {
      supervisionTimers.get(event.itemName).cancel();
      supervisionTimers.delete(event.itemName);
      console.log(event.itemName + ": Zustand hat sich geaendert, Ueberwachung wird abgebrochen");
    }
  }
});

File: patterns/mqtt-state-supervision/rules-python/mqtt_state_supervision.py
from openhab import rule, logger
from openhab.triggers import ItemCommandTrigger, ItemStateChangeTrigger

from supervision import supervise_state, cancel_supervision

T_RB_SECONDS = 30


def alert(item_name, expected_state):
    logger.warn(item_name + " hat den Zustand " + expected_state + " nicht erreicht")
    # z. B. Benachrichtigung senden


@rule(triggers=[ItemCommandTrigger("MySwitch")])
class MQTTStateSupervision:
    def execute(self, module, input):
        event = input.get("event")
        if not event:
            return

        expected_state = str(event.getItemCommand())
        item_name = event.getItemName()
        logger.info(item_name + ": Ueberwachung gestartet, erwarte " + expected_state + " innerhalb " + str(T_RB_SECONDS) + "s")
        supervise_state(item_name, expected_state, alert, T_RB_SECONDS)


@rule(triggers=[ItemStateChangeTrigger("MySwitch")])
class MQTTStateUpdate:
    def execute(self, module, input):
        event = input.get("event")
        if not event:
            return

        item_name = event.getItemName()
        logger.info(item_name + ": Zustand hat sich geaendert, Ueberwachung wird abgebrochen")
        cancel_supervision(item_name)
